src/libertem/io/dataset/memory.py

```diff
@@ -221,33 +229,51 @@ def __init__(self, tileshape=None, num_partitions=None, data=None, sig_dims=2,
         self.data = data
         self._base_shape = base_shape
         self.num_partitions = num_partitions
+
+        if sig_dims is None:
+            if sig_shape is not None:
+                sig_dims = len(sig_shape)
+            elif nav_shape is not None:
+                sig_dims = len(data.shape) - len(nav_shape)
+            else:
+                sig_dims = 2
+        else:
+            if sig_shape is not None and len(sig_shape) != sig_dims:
+                raise ValueError(
+                    f"Length of sig_shape {sig_shape} not matching sig_dims {sig_dims}."
+                )
+
         self.sig_dims = sig_dims
         if nav_shape is None:
-            self._nav_shape = self.shape.nav
+            nav_shape = data.shape[:-sig_dims]
         else:
-            self._nav_shape = tuple(nav_shape)
+            nav_shape = tuple(nav_shape)
         if sig_shape is None:
-            self._sig_shape = self.shape.sig
+            sig_shape = data.shape[-sig_dims:]
         else:
-            self._sig_shape = tuple(sig_shape)
-            self.sig_dims = len(self._sig_shape)
+            sig_shape = tuple(sig_shape)
+        reshaped_data = self.data.reshape((-1, *sig_shape))
```

Comment:
This reshape will fail if data is `scipy.sparse` or other 2d-only format, yet we supply a 2D sig_shape. The reshaped view is only used to calculate the image_count, so it is possible support this at least for creation of the dataset. As for generating tiles from this combination of 2D scipy.sparse => 3/4D datacube I don't yet know if that's supported?

Reply:
Nice catch! This particular instance is now replaced. Didn't test if scipy.sparse generally works, though.

Reply:
I looked into it and scipy.sparse matrices don't work with MemoryDataSet in general, only sparse.pydata.org arrays. Not sure if this should generate a different error by LiberTEM or if it should just fail naturally.